Suggest an appropriate module name with the 'defmodule' snippet

bottlenecked opened the pull request:
Hi all, this is a feature that I always wanted to have and now took some time to implement.

Basically when creating a new file and writing a `defmod...` prompt for auto-completion I wanted a likely module_name to be suggested along with completing the `defmodule do .. end` snippet.

This is what my attempt looks like

https://user-images.githubusercontent.com/622973/160127379-746dfd4a-05cc-4d3e-ad5e-a4c11ebe5152.mov

There are some limitations I'm aware of (and probably tens of others that I'm not aware of!)
- Nested module names: the users will be getting a suggestion based on the full file path instead of taking into account the parent module and basically doing nothing (let the users name their module)
- Phoenix conventions are not observed. e.g with controller names the suggested module name will be `MyProjectWeb.Controllers.PageController` instead of `MyProjectWeb.PageController`

Still, do let me know if this PR is on the right track and if so also let me know if these limitations (and possibly others) should be fixed before attempting to ship something like this to the users.

Thanks for taking the time to review!

sharpfun reviewed (commented):
Great staff, here are some ideas)

bottlenecked commented on apps/language_server/lib/language_server/providers/completion.ex:
```diff
@@ -593,6 +601,51 @@ defmodule ElixirLS.LanguageServer.Providers.Completion do
+
+      [file, "exs"] ->
+        if String.ends_with?(file, "_test") do
+          do_suggest_module_name(reversed_path, [file], topmost_parent: "test")
```
@sharpfun Thanks for the suggestions, but I'm hesitant to go any further at this point before anyone of the maintainers chimes in because it might be the wrong way to go about it

lukaszsamson replied:
We can't make this work for everyone. I'd aim for covering basic mix project and phoenix app as those are most commonly used.

lukaszsamson commented on apps/language_server/lib/language_server/providers/completion.ex:
```diff
@@ -576,6 +578,12 @@ defmodule ElixirLS.LanguageServer.Providers.Completion do
+  defp snippet_for({"Kernel", "defmodule"}, %{uri: uri}) when is_binary(uri) do
+    # In a mix project the uri can be something like "/some/code/path/project/lib/project/sub_path/myfile.ex"
+    # so we'll try to guess the appropriate module name from the path
+    "defmodule #{suggest_module_name(uri)}$1 do\n\t$0\nend"
```
convert URI to path. SourceFile has function for that

lukaszsamson commented on apps/language_server/lib/language_server/providers/completion.ex:
```diff
@@ -593,6 +601,51 @@ defmodule ElixirLS.LanguageServer.Providers.Completion do
+    do_suggest_module_name(rest, [dir_name | module_name_acc], opts)
+  end
+
+  defp do_suggest_module_name([], _module_name_acc, _opts) do
```
is it optimal to go all the way up to `/` or `C:\`? We could stop at workdir level.

lukaszsamson commented:
Nice feature @bottlenecked. I'd be happy to marge it. Can you verify if it works with umbrella apps and on windows? Those two are often tricky.

bottlenecked commented:
Thanks @lukaszsamson for the encouragement.

So I had some time to move forward with this to basically implement some of the suggestions here
- use `SourceFile.path_from_uri/1` to work with the path instead of the uri
- umbrellas look like they're already handled by the current approach so I added a test to prevent regressions in the future when used with umbrella apps
- Special cased the handling of known Phoenix folders directly under `*_web`

What I still haven't done is 
- make sure that this works on Windows (my feeling is that it will because `Path.split/1` should be able to handle Windows paths just fine but we'll see)
- handle files under the `test/support` folder. I'm not sure that there's a general rule that can be used like e.g. the top level module should always be `MyProjectWeb.*`, what do you think the rule should be like? (or maybe we just don't handle this case and we return the defmodule snippet without a suggested module name
- Stop traversing the path when reaching the project dir. Note: opening orphaned files from different parts of the disk into the same vscode instance might lead to unexpected behaviour, I'll see about testing that case

bottlenecked commented:
Some more thoughts on stopping at the project_dir level. It seems to me that the way the code is written, most of the times the traversal would stop at either the `lib` or the `test` folder levels. I guess I'd expect most projects to be structured like the mix new / mix phx.new structure the generated code, so maybe it would not be as bad to traverse the path all the way to the top when the projects are not structured as expected (should be few enough?). Plus, I think it would be pretty cheap to go all the way to the top anyway even for extreme cases of 1000 levels of deep folder nesting. What do you think @lukaszsamson?

bottlenecked commented:
I managed to test this on Windows (virtualbox VM on mac) and as ~hoped for~ expected `Path.split/1` does the right thing on Windows paths too

https://user-images.githubusercontent.com/622973/163556928-4cfd7e1e-de0b-466c-9c0d-6d5c232bd074.mov

bottlenecked commented:
I fixed the broken test (extra space at end of line was trimmed out) and removed the debug `IO.inspect` expression that previously added in d3b624f

noozo commented:
Much needed. Now i can get rid of the "Insert Elixir Module name here" extension. Genius!

jared-mackey commented:
After having used this for a long while now, I just wanted to chime in and say thanks again for this. It is such a great addition! 🙏🏻 ❤️
